librad/src/meta/user.rs

```diff
@@ -0,0 +1,215 @@
+// This file is part of radicle-link
+// <https://github.com/radicle-dev/radicle-link>
+//
+// Copyright (C) 2019-2020 The Radicle Team <dev@radicle.xyz>
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License version 3 or
+// later as published by the Free Software Foundation.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+use serde::{Deserialize, Serialize};
+use urltemplate::UrlTemplate;
+
+use crate::{
+    keys::device::PublicKey,
+    meta::{
+        common::{EmailAddr, Label, Url},
+        entity::{
+            data::{EntityBuilder, EntityData},
+            Entity,
+            Error,
+        },
+        profile::{Geo, ProfileImage, UserProfile},
+        serde_helpers,
+    },
+};
+
+#[derive(Clone, Deserialize, Serialize, Debug, Default, PartialEq)]
+pub struct UserInfo {
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub profile: Option<ProfileRef>,
+
+    #[serde(
+        skip_serializing_if = "Option::is_none",
+        serialize_with = "serde_helpers::urltemplate::serialize_opt",
+        deserialize_with = "serde_helpers::urltemplate::deserialize_opt"
+    )]
+    pub largefiles: Option<UrlTemplate>,
+}
+
+impl UserInfo {
+    pub fn new() -> Self {
+        Self {
+            profile: None,
+            largefiles: None,
+        }
+    }
+}
+
+pub type UserData = EntityData<UserInfo>;
+
+impl UserData {
+    fn profile_ref_mut(&mut self) -> Result<&mut ProfileRef, Error> {
+        match &mut self.info.profile {
+            Some(p) => Ok(p),
+            None => Err(Error::MissingRootHash),
```

Comment:
I don't understand this error in relation to this function :thinking: Same goes for below in `user_profile_mut`.

What's the `RootHash` have to do with the `ProfileRef` not being set?

Reply:
You are right, it makes no sense at all!
Sorry, this was part of a "FIXME" that got lost :-/
I was using a placeholder error to make it compile and I never implemented the real error.
Now I pushed a "BuilderError" for this case.